vllm/core/block/token_ids.py

```diff
@@ -0,0 +1,257 @@
+from typing import Iterable, List, NamedTuple, Optional, Tuple, overload
+
+
+class TokenRangeAnnotation(NamedTuple):
+    """
+    Annotates a range of placeholder tokens to capture content that will
+    replace them.
+    """
+
+    content_hash: int
+    content_offset: int
+    token_start_index: int
+    token_end_index: int
+
+    @property
+    def token_count(self) -> int:
+        return self.token_end_index - self.token_start_index
+
+    @staticmethod
+    def are_adjacent(left: "TokenRangeAnnotation",
+                     right: "TokenRangeAnnotation") -> bool:
+        """
+        Indicates whether two annotations represent adjacent ranges in the
+        hashed content.
+        """
+
+        return (left.content_hash == right.content_hash and
+                left.content_offset + left.token_count == right.content_offset)
+
+    def clipped_to_slice(self, tokens_start: int,
+                         tokens_end: int) -> Optional["TokenRangeAnnotation"]:
+        """
+        Computes a new TokenRangeAnnotation that corresponds to the same
+        content in a slice of the original token IDs.
+
+        For example, consider the following token IDs/annotations:
+
+        AAAA BBBB What do these images have in common?
+        
+        A = TokenRangeAnnotation(0xA, 0, 0, 4)
+        B = TokenRangeAnnotation(0xB, 0, 5, 9)
+
+        tokens = AAAA BBBB What do these images have in common?
+                [AAAA]
+
+        A.clipped_to_slice(0, 4) = TokenRangeAnnotation(0xA, 0, 0, 4)
+        B.clipped_to_slice(0, 4) = None
+
+        tokens = AAAA BBBB What do these images have in common?
+                  [AA BB]
+
+        A.clipped_to_slice(2, 7) = TokenRangeAnnotation(0xA, 2, 0, 2)
+        B.clipped_to_slice(2, 7) = TokenRangeAnnotation(0xB, 0, 3, 5)
+
+        tokens = AAAA BBBB What do these images have in common?
+                     [BBBB What]
+
+        A.clipped_to_slice(5, 14) = None
+        B.clipped_to_slice(5, 14) = TokenRangeAnnotation(0xB, 0, 0, 4)
+        """
+
+        unclamped_annotation_start = self.token_start_index - tokens_start
+        unclamped_annotation_end = self.token_end_index - tokens_start
+
+        annotation_start = max(0, unclamped_annotation_start)
+        annotation_end = min(tokens_end - tokens_start,
+                             unclamped_annotation_end)
+
+        if annotation_start >= annotation_end:
+            # There is no overlap.
+            return None
+
+        return TokenRangeAnnotation(content_hash=self.content_hash,
+                                    content_offset=self.content_offset +
+                                    annotation_start -
+                                    unclamped_annotation_start,
+                                    token_start_index=annotation_start,
+                                    token_end_index=annotation_end)
+
+
+class TokenIds:
+    token_ids: Tuple[int, ...]
```

Comment:
I don't particularly love `TokenIds.token_ids`. Maybe the type should just be`Tokens`? Maybe `BlockTokens`?